### DISCUSSION TITLE
Restricting Overlapping of two elements while dragging

### DISCUSSION THREAD

**Discussion by ROYopedia** [clientIO/joint#2289] [Q&A]
### Introduction

I hope you're doing well. I'm currently working on a project that involves preventing overlapping elements in a paper-based layout, and I'm facing a challenge that I believe the JointJS community could help me with. In this context, I'm utilizing the JointJS Planogram app demo, which incorporates the JointJS library along with a validator for this purpose.

In the code snippet provided below, you can see the relevant portion of the validator implementation:
```
const validator = new dia.Validator({
    commandManager,
    cancelInvalid: true,
});
validator.validate("change:position", validateChangePosition(graph));

```

```
export const validateChangePosition = (graph) => (err, command, next) => {
    console.log("Inside Validate Change Position");
    if (command.options.freeTransform) {
        // Position is changed after resize
        return next(err);
    }
    const cell = graph.getCell(command.data.id);
    if (!isPositionValid(graph, cell)) {
        return next({ cell, msg: "positioning error" });
    }
    return next(err);
};
export const isPositionValid = (graph, element) => {
    if (element.get("type") === "ngv.Node") {
        if (doesNodeOverlap(graph, element)) {
            return false;
        }
    } 
    return true;
};

export const doesNodeOverlap = (graph, shelf) => {
    return graph.findModelsUnderElement(shelf).length > 0;
};
```
However, I'm facing a significant issue with this approach:

The current implementation works well for cells that do not have any embedded cells.
When dealing with parent cells that contain embedded child cells, the parent cell's overlap prevention works as intended. However, the embedded child cells do not effectively restrict overlapping when the parent cell is being dragged.
In essence, I'm looking for assistance in adapting this code to ensure that when a parent cell (along with its embedded child cells) is dragged, the entire group should prevent overlap with other elements.

I'm eager to hear your suggestions, insights, and potential solutions to tackle this challenge effectively. Thank you for your time and expertise!

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS+

**Comment by jamesgeorgewilliams** [comment]
Hi, can you try and format the code with sensible spacing using code blocks? It is difficult to read at the moment. If other people are searching this discussion, it will be easier for them to read. Thanks

https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

https://github.com/clientIO/joint/wiki/Asking-Technical-Questions#3-include-steps-to-reproduce-problem

**Reply by ROYopedia** [reply]
Done

**Comment by kumilingus** [comment]
Good question. You should only validate the position of the container (not its children). Use the code below to determine which element is being dragged.

```js
const commandManager = new dia.CommandManager({
    graph,
    // tell the command manager to record also the `translateBy` option sent along with the position change
    // `translateBy` refers to an id of the element which is being dragged by the user
    applyOptionsList: ['translateBy']
});

const validator = new dia.Validator({ commandManager });
validator.validate('change:position', function(err, change) {
    if (change.options.translateBy === change.data.id) {
        // This is the topmost element which is being dragged.
    }
});
```

**Reply by ROYopedia** [reply]
I didn't fully understand your explanation.
Could you please provide a code reference or a small demo to help clarify the concept?

**Reply by kumilingus** [reply]
Is this more clear?
```js
const commandManager = new dia.CommandManager({
    graph,
    // tell the command manager to record also the `translateBy` option sent along with the position change
    // `translateBy` refers to an id of the element which is being dragged by the user
    applyOptionsList: ['translateBy']
});

const validator = new dia.Validator({ commandManager });
validator.validate('change:position', function(err, change, next) {
    if (change.options.translateBy === change.data.id) {
           const cell = graph.getCell(command.data.id);
            if (!isPositionValid(graph, cell)) {
                  return next({ cell, msg: "positioning error" });
            }
    }
    next(err);
});
```
---
All this code above was to make sure you don't check the position of the child elements, since they always overlap with their parent.

But you might also want to modify the `isPositionValid()` function to ignore all elements that are embedded in or ancestors of the element being dragged.
